jolt-core/src/jolt/trace/rv.rs

```diff
@@ -0,0 +1,721 @@
+use ark_ff::PrimeField;
+use enum_dispatch::enum_dispatch;
+use strum_macros::{EnumCount, EnumIter, FromRepr};
+
+use crate::jolt::instruction::add::ADD32Instruction;
+use crate::jolt::instruction::and::ANDInstruction;
+use crate::jolt::instruction::beq::BEQInstruction;
+use crate::jolt::instruction::bne::BNEInstruction;
+use crate::jolt::instruction::bge::BGEInstruction;
+use crate::jolt::instruction::bgeu::BGEUInstruction;
+use crate::jolt::instruction::bltu::BLTUInstruction;
+use crate::jolt::instruction::jal::JALInstruction;
+use crate::jolt::instruction::jalr::JALRInstruction;
+use crate::jolt::instruction::sltu::SLTUInstruction;
+use crate::jolt::instruction::or::ORInstruction;
+use crate::jolt::instruction::sll::SLLInstruction;
+use crate::jolt::instruction::slt::SLTInstruction;
+use crate::jolt::instruction::sra::SRAInstruction;
+use crate::jolt::instruction::srl::SRLInstruction;
+use crate::jolt::instruction::blt::BLTInstruction;
+use crate::jolt::instruction::xor::XORInstruction;
+use crate::jolt::instruction::{add::ADDInstruction, sub::SUBInstruction};
+use crate::jolt::instruction::{JoltInstruction, Opcode};
+use crate::jolt::subtable::LassoSubtable;
+use crate::jolt::vm::{pc::ELFRow, test_vm::TestInstructionSet};
+
+use super::{JoltProvableTrace, MemoryOp};
+
+// Reference: https://www.cs.sfu.ca/~ashriram/Courses/CS295/assets/notebooks/RISCV/RISCV_CARD.pdf
+#[derive(Debug, PartialEq, Eq, Clone, Copy, FromRepr)]
+#[repr(u8)]
+pub enum RV32IM {
+  ADD,
+  SUB,
+  XOR,
+  OR,
+  AND,
+  SLL,
+  SRL,
+  SRA,
+  SLT,
+  SLTU,
+  ADDI,
+  XORI,
+  ORI,
+  ANDI,
+  SLLI,
+  SRLI,
+  SRAI,
+  SLTI,
+  SLTIU,
+  LB,
+  LH,
+  LW,
+  LBU,
+  LHU,
+  SB,
+  SH,
+  SW,
+  BEQ,
+  BNE,
+  BLT,
+  BGE,
+  BLTU,
+  BGEU,
+  JAL,
+  JALR,
+  LUI,
+  AUIPC,
+  ECALL,
+  EBREAK,
+  MUL,
+  MULH,
+  MULSU,
+  MULU,
+  DIV,
+  DIVU,
+  REM,
+  REMU,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum RV32InstructionFormat {
+  R,
+  I,
+  S,
+  SB,
+  U,
+  UJ,
+}
```

Comment:
Instead of having all the Options on `RVTraceRow`, why not change these enum variants to structs?
```suggestion
struct RInstruction {
    pc: u64;
    opcode: RV32IM;
    rd: u64;
    rs1: u64;
    rs2: u64;
    rd_pre_val: u64;
    rd_post_val: u64;
    rs1_val: u64;
    rs2_val: u64;
}
// etc...
pub enum RV32InstructionFormat {
  R(RInstruction),
  I(IInstruction),
  S(SInstruction),
  SB(SBInstruction),
  U(UInstruction),
  UJ(UJInstruction),
}
```

Reply:
I went back and forth on this. If you look at the match statements below, we mostly don't want to match on the InstructionFormat which would add unnecessary unwrapping in most cases.

It's a bit more memory efficient, and slightly cleaner in specific cases (R instructions for example), but at the cost of readability.